bundles/azure/aso/psql/bundle/config/01-flexible-server.ytt.yml

```diff
@@ -0,0 +1,247 @@
+#@ load("@ytt:data", "data")
+#@ load("@ytt:assert", "assert")
+#@ load("@ytt:regexp", "regexp")
+#@ load("@ytt:ip", "ip")
+
+#@ def tags(resource_tags=[]):
+#@  tags_list = []
+#@  tags_list.extend(data.values.global_tags)
+#@  tags_list.extend(resource_tags)
+#@  tags_map = { x["key"]: x["value"] for x in tags_list if x["key"] != "" and x["value"] != "" }
+#@  return(tags_map)
+#@ end
+
+#@ flexibleServerK8SName    = data.values.name
+#@ dbSecretName          = data.values.name
+#@ resourceGroupK8SName  = data.values.name
+#@ serviceAccountName      = data.values.name + "-reader"
+#@ serviceAccountNamespace = data.values.namespace
+#@ roleName                = data.values.name + "-reading"
+
+#@ firewallRules = data.values.firewall_rules
+#@ if len(firewallRules) == 0:
+#@   firewallRules = [{"startIpAddress": "0.0.0.0", "endIpAddress": "0.0.0.0", "tags": []}]
```

Comment:
_nit/question_: If you want to go with that as a default, why don't you make that the default in the values-schema? It would then be exposed in the openAPI  schema and thus discoverable by clients (e.g. the tanzu CLI).

Reply:
Because a default for a list means that the default value will ALWAYS be in the final list, no matter the value we supply.
What I expect from a default value is that it is used only if I don't provide anything, which is the case with strings, numbers and booleans, but lists and maps behave differently, as the default is merged to the value supplied by the user. It might make sense for maps (if I don't provide a field of the map I will get its default value) but certainly it does not for lists, as I don't want the default value to be used at all if I provide an actual value.

Reply:
Huh! Didn't realize that, but `ytt` behaves differently when you set the default via annotation as opposed to inline: 

<details><summary>default inline</summary>

```console
$ grep -R \^ f input_data_values.yml
f/thing.yaml:#@ load("@ytt:data","data")
f/thing.yaml:---
f/thing.yaml:firewall_stuff: #@ data.values.firewall_rules
f/schema.yml:#@data/values-schema
f/schema.yml:---
f/schema.yml:firewall_rules:
f/schema.yml:- startIpAddress: "1.1.1.1"
f/schema.yml:  endIpAddress: "2.2.2.2"
input_data_values.yml:#@data/values
input_data_values.yml:---
input_data_values.yml:firewall_rules:
input_data_values.yml:- startIpAddress: 5.5.5.5
input_data_values.yml:  endIpAddress: 6.6.6.6
$ ytt -f f
firewall_stuff: []
$ ytt -f f -f input_data_values.yml
firewall_stuff:
- startIpAddress: 5.5.5.5
  endIpAddress: 6.6.6.6
$ ytt -f f --data-values-schema-inspect -o openapi-v3
openapi: 3.0.0
info:
  version: 0.1.0
  title: Schema for data values, generated by ytt
paths: {}
components:
  schemas:
    dataValues:
      type: object
      additionalProperties: false
      properties:
        firewall_rules:
          type: array
          items:
            type: object
            additionalProperties: false
            properties:
              startIpAddress:
                type: string
                default: 1.1.1.1
              endIpAddress:
                type: string
                default: 2.2.2.2
          default: []
$
```

</details>

<details><summary>default from annotation</summary>

```console
$ grep -R \^ f input_data_values.yml
f/thing.yaml:#@ load("@ytt:data","data")
f/thing.yaml:---
f/thing.yaml:firewall_stuff: #@ data.values.firewall_rules
f/schema.yml:#@data/values-schema
f/schema.yml:---
f/schema.yml:#@schema/default [{"startIpAddress":"1.1.1.1", "endIpAddress":"2.2.2.2"}]
f/schema.yml:firewall_rules:
f/schema.yml:- startIpAddress: ""
f/schema.yml:  endIpAddress: ""
input_data_values.yml:#@data/values
input_data_values.yml:---
input_data_values.yml:firewall_rules:
input_data_values.yml:- startIpAddress: 5.5.5.5
input_data_values.yml:  endIpAddress: 6.6.6.6
$ ytt -f f
firewall_stuff:
- startIpAddress: 1.1.1.1
  endIpAddress: 2.2.2.2
$ ytt -f f -f input_data_values.yml
firewall_stuff:
- startIpAddress: 1.1.1.1
  endIpAddress: 2.2.2.2
- startIpAddress: 5.5.5.5
  endIpAddress: 6.6.6.6
$ ytt -f f --data-values-schema-inspect -o openapi-v3
openapi: 3.0.0
info:
  version: 0.1.0
  title: Schema for data values, generated by ytt
paths: {}
components:
  schemas:
    dataValues:
      type: object
      additionalProperties: false
      properties:
        firewall_rules:
          type: array
          items:
            type: object
            additionalProperties: false
            properties:
              startIpAddress:
                type: string
                default: ""
              endIpAddress:
                type: string
                default: ""
          default:
          - startIpAddress: 1.1.1.1
            endIpAddress: 2.2.2.2
$
```

</details>

And neither does what I would have expected :)

Reply:
https://carvel.dev/ytt/#gist:https://gist.github.com/hoegaarden/1fcb5d5d7c2afd7645342943992de52b